### ISSUE TITLE
Class.getAnnotations() returns invalid results

### ISSUE THREAD

**Issue by koutheir** [oracle/graal#4524]
Using `native-image`, calling `Class.getAnnotations()` returns invalid results.

# Sample to reproduce
```java
import java.lang.annotation.Annotation;
import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

@Retention(RetentionPolicy.RUNTIME)
@Repeatable(ContainerAnnot.class)
@interface RepeatMeAnnot {}

@Retention(RetentionPolicy.RUNTIME)
@interface DifferentAnnot {}

@Retention(RetentionPolicy.RUNTIME)
@interface ContainerAnnot {
    RepeatMeAnnot[] value();
}

@DifferentAnnot @RepeatMeAnnot @RepeatMeAnnot @RepeatMeAnnot @RepeatMeAnnot
class Test {}

public class Hello {
    public static void main(String argv[]) {
        System.exit(run());
    }

    public static int run() {
        try {
            Class<Test> cl = Test.class;
            Annotation[] annotations = cl.getAnnotations();
            if (annotations == null) {
                System.err.println("[ERROR] cl.getAnnotations() == null");
                return 2;
            }
            if (annotations.length != 2) {
                System.err.println("[ERROR] cl.getAnnotations().length == " + annotations.length);
                return 3;
            }

            System.err.println("[OK] Done.");
            return 0;
        } catch (Exception ex) {
            return 1;
        }
    }
}
```

# Expected output
```bash
$ java Hello
[OK] Done.
```

# Actual output
```bash
$ ./hello
[ERROR] cl.getAnnotations().length == 0
```

# Compilation output
```bash
$ <...>/bin/native-image --native-image-info --verbose Hello
Executing [
<...>/bin/java \
-XX:+UseParallelGC \
-XX:+UnlockExperimentalVMOptions \
-XX:+EnableJVMCI \
-Dtruffle.TrustAllTruffleRuntimeProviders=true \
-Dtruffle.TruffleRuntime=com.oracle.truffle.api.impl.DefaultTruffleRuntime \
-Dgraalvm.ForcePolyglotInvalid=true \
-Dgraalvm.locatorDisabled=true \
-Dsubstratevm.IgnoreGraalVersionCheck=true \
--add-exports=java.base/com.sun.crypto.provider=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.event=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.loader=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.logger=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.misc=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.module=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.org.objectweb.asm=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.org.xml.sax.helpers=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.perf=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.ref=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.reflect=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.util.xml.impl=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.util.xml=ALL-UNNAMED \
--add-exports=java.base/sun.invoke.util=ALL-UNNAMED \
--add-exports=java.base/sun.net=ALL-UNNAMED \
--add-exports=java.base/sun.nio.ch=ALL-UNNAMED \
--add-exports=java.base/sun.reflect.annotation=ALL-UNNAMED \
--add-exports=java.base/sun.reflect.generics.reflectiveObjects=ALL-UNNAMED \
--add-exports=java.base/sun.reflect.generics.repository=ALL-UNNAMED \
--add-exports=java.base/sun.reflect.generics.tree=ALL-UNNAMED \
--add-exports=java.base/sun.security.jca=ALL-UNNAMED \
--add-exports=java.base/sun.security.provider=ALL-UNNAMED \
--add-exports=java.base/sun.security.ssl=ALL-UNNAMED \
--add-exports=java.base/sun.security.util=ALL-UNNAMED \
--add-exports=java.base/sun.text.spi=ALL-UNNAMED \
--add-exports=java.base/sun.util.calendar=ALL-UNNAMED \
--add-exports=java.base/sun.util.locale.provider=ALL-UNNAMED \
--add-exports=java.base/sun.util.resources=ALL-UNNAMED \
--add-exports=java.management/sun.management=ALL-UNNAMED \
--add-exports=java.xml.crypto/org.jcp.xml.dsig.internal.dom=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.aarch64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.amd64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.armv7a=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code.site=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code.stack=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.common=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.aarch64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.amd64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.armv7a=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.sparc=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.meta=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.runtime=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.services=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.sparc=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.events=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.internal.consumer=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.internal.handlers=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.internal.jfc=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.internal=ALL-UNNAMED \
--add-exports=org.graalvm.truffle/com.oracle.truffle.api.object=ALL-UNNAMED \
-XX:-UseJVMCICompiler \
-Xss10m \
-Xms1g \
-Xmx14g \
-Djava.awt.headless=true \
-Dorg.graalvm.version=dev \
-Dorg.graalvm.config=CE \
-Dcom.oracle.graalvm.isaot=true \
-Djava.system.class.loader=com.oracle.svm.hosted.NativeImageSystemClassLoader \
-Xshare:off \
-Djdk.internal.lambda.disableEagerInitialization=true \
-Djdk.internal.lambda.eagerlyInitialize=false \
-Djava.lang.invoke.InnerClassLambdaMetafactory.initializeLambdas=false \
-javaagent:<...>/lib/svm/builder/svm.jar \
-cp \
<...>/lib/svm/builder/svm-enterprise.jar:<...>/lib/svm/builder/javacpp-shadowed.jar:<...>/lib/svm/builder/llvm-wrapper-shadowed.jar:<...>/lib/svm/builder/objectfile.jar:<...>/lib/svm/builder/javacpp-platform-specific-shadowed.jar:<...>/lib/svm/builder/native-image-base.jar:<...>/lib/svm/builder/svm-llvm.jar:<...>/lib/svm/builder/svm-enterprise-llvm.jar:<...>/lib/svm/builder/pointsto.jar:<...>/lib/svm/builder/llvm-platform-specific-shadowed.jar:<...>/lib/svm/builder/svm.jar \
--module-path \
<...>/lib/truffle/truffle-api.jar \
'com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus' \
-imagecp \
<...>/playground:<...>/lib/svm/library-support.jar \
-H:Path=<...>/playground \
-H:+DumpTargetInfo \
-H:CLibraryPath=<...>/lib/svm/clibraries/linux-amd64 \
'-H:Class@explicit main-class=Hello' \
'-H:Name@main-class lower case as image name=hello'
]
========================================================================================================================
GraalVM Native Image: Generating 'hello'...
========================================================================================================================
[1/7] Initializing...                                                                                    (3.8s @ 0.07GB)
 Version info: 'GraalVM dev Java 11 EE'
# Printing compilation-target information to: <...>/playground/reports/target_info_20220429_104513.txt
[2/7] Performing analysis...  [******]                                                                   (3.4s @ 0.64GB)
# Printing native-library information to: <...>/playground/reports/native_library_info_20220429_104517.txt
   1,659 (71.42%) of  2,323 classes reachable
   1,642 (55.23%) of  2,973 fields reachable
   6,637 (63.49%) of 10,453 methods reachable
      20 classes,     0 fields, and   128 methods registered for reflection
      48 classes,    32 fields, and    47 methods registered for JNI access
[3/7] Building universe...                                                                               (0.5s @ 0.28GB)
[4/7] Parsing methods...      [*]                                                                        (0.2s @ 0.44GB)
[5/7] Inlining methods...     [****]                                                                     (0.3s @ 0.78GB)
[6/7] Compiling methods...    [***]                                                                      (5.1s @ 2.61GB)
[7/7] Creating image...                                                                                  (0.5s @ 2.76GB)
   2.44MB (46.90%) for code area:    3,137 compilation units
   2.27MB (43.59%) for image heap:     840 classes and 36,808 objects
 507.08KB ( 9.51%) for other data
   5.21MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
 330.54KB java.lang                                          632.30KB byte[] for general heap data
 239.57KB java.util                                          276.98KB byte[] for java.lang.String
 194.53KB com.oracle.svm.jni                                 238.18KB java.lang.Class
 128.84KB java.util.concurrent                               216.30KB java.lang.String
 113.21KB com.oracle.svm.core.reflect                         98.13KB java.util.HashMap$Node[]
 111.91KB java.math                                           96.06KB java.util.HashMap$Node
 107.86KB com.oracle.svm.core.genscavenge                     90.64KB char[]
  83.59KB com.oracle.svm.jni.functions                        48.19KB java.util.HashMap
  81.07KB java.util.stream                                    42.48KB byte[] for method metadata
  79.08KB java.io                                             39.85KB int[]
      ... 86 additional packages                                  ... 429 additional object types
                                           (use GraalVM Dashboard to see all)
------------------------------------------------------------------------------------------------------------------------
                        0.8s (5.3% of total time) in 16 GCs | Peak RSS: 3.38GB | CPU load: 11.55
------------------------------------------------------------------------------------------------------------------------
Produced artifacts:
 <...>/playground/hello (executable)
 <...>/playground/hello.build_artifacts.txt
========================================================================================================================
Finished generating 'hello' in 14.4s.
```

# GraalVM and my environment:
 - GraalVM version: EE 22.0.0.2, jvmci-22.0-b05
 - JDK major version: 11.0.14
 - OS: Linux
 - Architecture: amd64

**Comment by loicottet** [comment]
The behaviour you are witnessing was the intended behaviour of the 22.0 version, which was to only include annotations in the image if their type was actually accessed at runtime. This was changed for 22.1 to include all annotations regardless of reachability, and I was unable to reproduce this issue on the latest 22.1 version. Can you please confirm that this is the case?

**Comment by koutheir** [comment]
If this was fixed on version `22.1`, then I'll simply wait until we receive that version, and test this behavior again with that version.

**Comment by gilles-duboscq** [comment]
@koutheir, is this now resolved with 22.1 or do you still have issues with this?

**Comment by koutheir** [comment]
I confirm that this behavior was fixed with `native-image` version `22.1`.
